### DISCUSSION TITLE
Parameterizable global library written in Simatic AX?

### DISCUSSION THREAD

**Discussion by ahuca** [simatic-ax/.discussions#23] [Q&A]
Hello,

In the case of a "traditional" global library developed using the TIA Portal, one can package a tag table containing the user constants along with the library itself. Then TIA Portal project that reuse this library can grab the tag table from the library itself and adjust the user constants. This provides a mean to parameterize a global library.

Is there a way to achieve this functionality with an AX-developed global library?

**Comment by sjuergen** [comment]
Hi @ahuca, 

I would also like to understand, what's your use case behind your question.

As @PTKu mentioned, it is allowed to put named values and classes with public members into a library.

Regarding global tags within a library, I would like to share some thoughts: 
- it can't be avoided to use the same tag (e.g. myTag) in two different libraries. Since global tags can not be located within a namespace, it doesn't make sense for me.
- libraries should be independent and encapsulated. Change some external resources (in this case a global tag) should have impact on the behavior of the library. 
- constants should be constant. Constants should be adjustable. Otherwise it's a variable.
- In case of different sizes of arrays, you can use Array* - but adjust the size in your application and not in the lib.

**Reply by ahuca** [reply]
Hi @sjuergen, and thank you for your comment.

My current use case is exactly your last-mentioned bullet point, variable length array. I have an object A which has a number of B structs. The B-array size is dictated on the application side.

With array*, it can only be used in these cases (AX documentation).

> A variable-length array can ONLY be declared in the following variable sections:
> 
> - input, output or in-out section of methods and functions
> - in-out section of function block.

Which is understandable syntactically.

However, in this case, A has a "has-a" relationship with B, thus I would have liked to have a global constant which can be updated on the application side. For me, keeping B inside A is also a kind of encapsulation. Passing B-array into A from the application does not guarantee that elsewhere in the application B-array is not modified unexpectedly.

Though, I understand the limitation now that you mentioned that global tag does not have a namespace.

The best-case scenario would have been a dynamic memory reservation that can be run once every time the PLC boots up. Something that can be done with the [`__NEW`](https://help.codesys.com/api-content/2/codesys/3.5.15.0/en/_cds_operator_new/) operator and [`FB_DynMem_Buffer`](https://infosys.beckhoff.com/english.php?content=../content/1033/tcplclib_tc3_dynamicmemory/7156497547.html&id=7020316734011269815), which allowed me to create a [list implementation](https://github.com/ahuca/TcBase/blob/main/TcBase/TcBase/TcBase/Collection/FB_List.TcPOU) in TwinCAT 3.

I've been looking into an alternative using `REF_TO Array*`. So in an application, I could point the reference to an array in there. However I'm unsure yet how can I implement the actual "pointing" because REF_TO is only allowed in Temp variable in a FB, when my goal is to wrap the object A itself into a FB.
